mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp

```diff
@@ -1818,6 +1818,11 @@ struct RemoveOutsDependency : public OpRewritePattern<GenericOp> {
         if (sparse_tensor::getSparseTensorEncoding(operandVal.getType()))
           continue;
 
+        // If outs is wired from a block argument, keep the dependency to
+        // prevent the argument from being optimized away.
+        if (isa<BlockArgument>(operandVal))
```

Comment:
We can do something slightly better here. Check if this value is coming from a `DestinationStyleOpInterface` . If so replace it with the corresponding `init` value of that operation. I you do that you should be able to remove lines 1821 to 1835.

Reply:
We would still need to keep the check since the value can come from an operation that does not have the interface. The case I am trying to solve is a function with `iree.abi.output` on a function parameter.

Reply:
Yeah I understand. What I am saying is that this is being "too aggressive". This was added before destination passing style. What this pattern should be doing is that for elementwise operation, if outs is coming from a destination passing style op, then you replace the operand with the corresponding destination operand of the `outs`. In all other cases, it should just "do nothing". So it will also encompass your change, but is "more correct". Currently it is probably changing the semantics of the program too much.

Reply:
I believe we the community needs more discussion about this and make a clear specification how to use a tensor as output. That's going to be lots of work. For now,  the path using `iree.abi.output` and leaving the dependency as is can be a good start to at least have a working flow. 

Since the topic can have multiple different design choices and you have a different scope of concern, it would be better to take the concern and discussion outside of this PR and populate solutions there.

We had an offline discussion and Mahesh will create a separate issue to describe his concern and provide more information. Thanks for the help!